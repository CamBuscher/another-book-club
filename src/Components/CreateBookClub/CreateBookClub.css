.createClubForm {
  margin: auto;
}

.createClubForm input {
  width: 350px;
}

.createClubForm hr {
  width: 50%;
  margin-bottom: 20px;
}

.createClubForm h2 {
  color: whitesmoke;
  text-shadow: 2px 2px 4px black;
}