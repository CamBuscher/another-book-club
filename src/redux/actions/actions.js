export const setUser = user => ({
  type: 'SET_USER',
  user
});

export const updateBookClubs = (club, clubName) => ({
  type: 'UPDATE_BOOK_CLUBS',
  club,
  clubName
});

export const removeBookClub = (clubName) => ({
  type: 'REMOVE_CLUB',
  clubName
});

export const setCurrentClub = club => ({
  type: 'SET_CURRENT_CLUB',
  club
});

export const addBookToClub = club => ({
  type: 'ADD_BOOK_TO_CLUB',
  club
});