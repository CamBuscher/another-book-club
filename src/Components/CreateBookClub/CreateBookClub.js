import React, { Component } from 'react';
import uniqid from 'uniqid';
import { db } from '../../firebase';
import { connect } from 'react-redux';
import { updateBookClubs } from '../../redux/actions/actions';
import './CreateBookClub.css';
import PropTypes from 'prop-types';

export class CreateBookClub extends Component {
  constructor(props) {
    super(props);

    this.state = {
      clubName: '',
      description: ''
    };
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const id = uniqid();
    const { clubName, description } = this.state;
    const { user } = this.props;
    const newClub = {
      id, 
      clubName 
    };
    db.doCreateBookClub(id, clubName, [{...user}], description);
    db.updateUserBookClubs(user, newClub);
    this.props.updateBookClubs({[clubName]: {clubName, id, description}}, clubName);
  }

  byPropKey = (propertyName, value) => () => ({
    [propertyName]: value
  });

  render() {
    const { clubName, description } = this.state

    return (
      <form onSubmit={this.handleSubmit} className='createClubForm'> 
        <h2>Make a new book club</h2>
        <input 
          value={clubName}
          onChange={event => this.setState(this.byPropKey('clubName', event.target.value))}
          type="text"
          placeholder="What would you like to name your book club?"
        />
        <br />
        <input
          value={description}
          onChange={event => this.setState(this.byPropKey('description', event.target.value))}
          type="text"
          placeholder="Describe your book club?"
        />
        <br />
        <button type='submit'>Create </ button>
        <hr />
      </form>
    )
  }
}

CreateBookClub.propTypes = {
  user: PropTypes.object,
  updateBookClubs: PropTypes.func
};

export const mapStateToProps = ({user}) => ({ user });

export const mapDispatchToProps = dispatch => ({
  updateBookClubs: (club, clubName) => dispatch(updateBookClubs(club, clubName))
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateBookClub);