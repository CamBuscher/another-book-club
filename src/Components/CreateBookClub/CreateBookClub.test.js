import React from 'react';
import {updateBookClubs} from '../../redux/actions/actions';
import {CreateBookClub, mapDispatchToProps, mapStateToProps} from './CreateBookClub';
import {shallow} from 'enzyme';

describe('CreateBookClub', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(<CreateBookClub updateBookClubs={jest.fn()}/>);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls dispatch with a updateBookClubs action', () => {
    const dispatch = jest.fn();

    const actionToDispatch = updateBookClubs();

    const mappedProps = mapDispatchToProps(dispatch);

    mappedProps.updateBookClubs();

    expect(dispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('should return the user', () => {
    const mockState = {
      user: {
        bookClubs: [],
        id: 'asdasd'
      }
    };

    let expected = { bookClubs: [], id: 'asdasd' };

    const mappedProps = mapStateToProps(mockState).user;

    expect(mappedProps).toEqual(expected);
  });

  it('should update state when passed correct propKey', () => {
    wrapper.instance().byPropKey('description', 'hello');
    expect((wrapper.state()).description).toEqual('');
  });
});