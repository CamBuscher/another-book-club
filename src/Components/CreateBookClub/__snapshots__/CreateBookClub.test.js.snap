// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBookClub should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateBookClub
    updateBookClubs={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Make a new book club
</h2>,
        <input
          onChange={[Function]}
          placeholder="What would you like to name your book club?"
          type="text"
          value=""
/>,
        <br />,
        <input
          onChange={[Function]}
          placeholder="Describe your book club?"
          type="text"
          value=""
/>,
        <br />,
        <button
          type="submit"
>
          Create 
</button>,
        <hr />,
      ],
      "className": "createClubForm",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Make a new book club",
        },
        "ref": null,
        "rendered": "Make a new book club",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "What would you like to name your book club?",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Describe your book club?",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Create ",
          "type": "submit",
        },
        "ref": null,
        "rendered": "Create ",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Make a new book club
</h2>,
          <input
            onChange={[Function]}
            placeholder="What would you like to name your book club?"
            type="text"
            value=""
/>,
          <br />,
          <input
            onChange={[Function]}
            placeholder="Describe your book club?"
            type="text"
            value=""
/>,
          <br />,
          <button
            type="submit"
>
            Create 
</button>,
          <hr />,
        ],
        "className": "createClubForm",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Make a new book club",
          },
          "ref": null,
          "rendered": "Make a new book club",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "What would you like to name your book club?",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Describe your book club?",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Create ",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Create ",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
